lib/SILOptimizer/Utils/LexicalDestroyFolding.cpp

```diff
@@ -0,0 +1,687 @@
+        if (!BorrowingOperand(use).visitScopeEndingUses([&](Operand *end) {
+              return end->getOperandOwnership() != OperandOwnership::Reborrow;
+            })) {
+          return false;
+        }
+        if (auto *bbi = dyn_cast<BeginBorrowInst>(use->getUser())) {
+          worklist.insert(bbi);
+        }
+      }
+      // We are only interested in uses that occur after %introducer--those are
+      // the uses which we want to guarantee are not inside the liveness
+      // boundary of introducer.
+      if (!dominanceTree.dominates(introducer, use->getUser()))
+        continue;
+      originalBorroweeUses.push_back(use);
+    }
+  }
+  return true;
+}
+
+bool LexicalDestroyFolding::originalBorroweeHasUsesWithinBorrowScope() {
+  PrunedLiveness liveness;
+  borrowedValue.computeLiveness(liveness);
+  DeadEndBlocks deadEndBlocks(function);
+  if (!findDominatedTransitiveUsesOfOriginalBorrowee())
+    return false;
+  return !liveness.areUsesOutsideBoundary(originalBorroweeUses, &deadEndBlocks);
+}
+
+//===----------------------------------------------------------------------===//
+//                             MARK: Predicates
+//===----------------------------------------------------------------------===//
+
+size_t LexicalDestroyFolding::extendedSimpleUseCount(ApplyInst *ai) const {
+  return extendedSimpleUseCountPerApplySite.lookup(ai);
+}
+
+Optional<LexicalDestroyFolding::Match>
+LexicalDestroyFolding::definesMatchingInstructionSequence(
+    SILInstruction *inst) const {
+  // Look specifically for
+  //
+  //   apply
+  //   end_borrow // inst
+  //   ...
+  //   destroy_value %originalBorrowee
+  //
+  // TODO: Also match try_apply and begin_apply all of the instructions
+  //       immediately following which are end_borrows.
+  auto *ebi = dyn_cast<EndBorrowInst>(inst);
+  if (!ebi)
+    return None;
+  auto *dvi =
+      beginsOriginalBorroweeDestroyingSequence(ebi->getNextInstruction());
+  if (!dvi)
+    return None;
+  auto *ai = dyn_cast_or_null<ApplyInst>(ebi->getPreviousInstruction());
+  if (!ai)
+    return {{nullptr, ebi, dvi}};
+
+  return {{ai, ebi, dvi}};
+}
+
+DestroyValueInst *
+LexicalDestroyFolding::beginsOriginalBorroweeDestroyingSequence(
+    SILInstruction *dvi) const {
+  SILInstruction *inst = dvi;
+  while (inst) {
+    if (!CanonicalizeOSSALifetime::mustNotHoistDestroyOver(inst->getKind())) {
+      // This is not an instruction that CanonicalOSSALifetime would not hoist a
+      // destroy above.  In other words, CanonicalOSSALifetime would have
+      // hoisted
+      //     destroy_value %original_borrowee
+      // over this instruction if it could have.  Stop looking.
+      return nullptr;
+    }
+    if (auto *dvi = dyn_cast<DestroyValueInst>(inst)) {
+      if (dvi->getOperand() == originalBorrowee) {
+        return dvi;
+      }
+    }
+    for (auto &operand : inst->getAllOperands()) {
+      if (operand.get() == originalBorrowee) {
```

Comment:
shouldn't this be an assert?

Reply:
No, when CanonicalizeOSSALifetime puts destroys after last uses, it won't hoist over some instructions like `destroy_value %other` in order to avoid churning while looping.